### PR TITLE
Network update

### PR DESCRIPTION
**Pull request by eugenevinitsky**
- put in correct version of i210
- put in correct version of straight road
- better logging in train.py
- curricula in i210 env

### CODE REVIEW

File: flow/controllers/velocity_controllers.py [hunk 1]
```diff
@@ -45,6 +48,10 @@ def __init__(self,
         self.d_2 = 1.0
         self.d_3 = 0.5
 
+        self.danger_edges = danger_edges if danger_edges else {}
```
Comment (kjang96) [inline flow/controllers/velocity_controllers.py:51]: It'd be easier to set `danger_edges={}` in the function header, turning {} into the default vlaue

File: flow/core/rewards.py [hunk 2]
```diff
@@ -389,7 +389,7 @@ def miles_per_megajoule(env, veh_ids=None, gain=.001):
         speed = env.k.vehicle.get_speed(veh_id)
         # convert to be positive since the function called is a penalty
         power = -vehicle_energy_consumption(env, veh_id, gain=1.0)
-        if power > 0 and speed >= 0.0:
+        if power > 0 and not speed < 0.1:
```
Comment (kjang96) [inline flow/core/rewards.py:392]: speed >= 0.1

File: flow/envs/multiagent/base.py [hunk 3]
```diff
@@ -4,7 +4,7 @@
 import numpy as np
 import random
 import traceback
-from gym.spaces import Box
+from gym.spaces import Box, Dict
```
Comment (kjang96) [inline flow/envs/multiagent/base.py:7]: Dict import not used, remove

File: flow/envs/test.py [hunk 4]
```diff
@@ -52,3 +52,17 @@ def compute_reward(self, rl_actions, **kwargs):
...
         """See class definition."""
         return np.array([])
+
+class TestI210Env(TestEnv):
+
+    def additional_command(self):
+        edge = "119257908#0"
+        edge_length = self.k.network.edge_length(edge)
+        for veh_id in self.k.vehicle.get_ids():
+            edge = self.k.vehicle.get_edge(veh_id)
```
Comment (kjang96) [inline flow/envs/test.py:62]: Need new variable for this lil guy or there will be a name conflict and then line 66 makes no sense

File: flow/visualize/visualizer_rllib.py [hunk 5]
```diff
@@ -160,6 +169,11 @@ def visualizer_rllib(args):
     else:
         env = gym.make(env_name)
 
+    if hasattr(env, "reroute_on_exit"):
+        env.reroute_on_exit = False
```
Comment (kjang96) [inline flow/visualize/visualizer_rllib.py:173]: Shouldn't this be True?

File: flow/envs/multiagent/i210.py [hunk 6]
```diff
@@ -109,38 +140,48 @@ def action_space(self):
     def _apply_rl_actions(self, rl_actions):
         """See class definition."""
         # in the warmup steps, rl_actions is None
+        id_list = []
+        accel_list = []
         if rl_actions:
+            t = time()
```
Comment (kjang96) [inline flow/envs/multiagent/i210.py:146]: Remove

File: flow/envs/multiagent/i210.py [hunk 7]
```diff
@@ -109,38 +140,48 @@ def action_space(self):
...
-                if lead_id in ["", None]:
-                    # in case leader is not visible
-                    lead_speed = SPEED_SCALE
-                    headway = HEADWAY_SCALE
-                else:
-                    lead_speed = self.k.vehicle.get_speed(lead_id)
-                    headway = self.k.vehicle.get_headway(rl_id)
-                veh_info.update({rl_id: np.array([speed / SPEED_SCALE, headway / HEADWAY_SCALE,
-                                                  lead_speed / SPEED_SCALE])})
+                if (self.control_range and self.k.vehicle.get_x_by_id(rl_id) < self.control_range[1] \
```
Comment (kjang96) [inline flow/envs/multiagent/i210.py:165]: I'd suggest making a in_control_range function to simplify this here and in the other places with the same logic (I think mostly in this file)

File: flow/envs/multiagent/i210.py [hunk 8]
```diff
@@ -109,38 +140,48 @@ def action_space(self):
...
+                    else:
+                        lead_speed = self.k.vehicle.get_speed(lead_id)
+                        headway = self.k.vehicle.get_headway(rl_id)
+                    veh_info.update({rl_id: np.array([speed / SPEED_SCALE, headway / HEADWAY_SCALE,
+                                                      lead_speed / SPEED_SCALE])})
         else:
             veh_info = {rl_id: np.concatenate((self.state_util(rl_id),
                                                self.veh_statistics(rl_id)))
                         for rl_id in self.k.vehicle.get_rl_ids()}
+        # print('time to get state is ', time() - t)
```
Comment (kjang96) [inline flow/envs/multiagent/i210.py:184]: Remove

File: flow/envs/multiagent/i210.py [hunk 9]
```diff
@@ -149,28 +190,115 @@ def compute_reward(self, rl_actions, **kwargs):
         if rl_actions is None:
             return {}
 
+        t = time()
```
Comment (kjang96) [inline flow/envs/multiagent/i210.py:193]: Remove

File: flow/envs/multiagent/i210.py [hunk 10]
```diff
@@ -200,28 +329,44 @@ def additional_command(self):
...
+                        break
 
             departed_ids = self.k.vehicle.get_departed_ids()
-            if len(departed_ids) > 0:
+            if isinstance(departed_ids, tuple) and len(departed_ids) > 0:
                 for veh_id in departed_ids:
                     if veh_id not in self.observed_ids:
                         self.k.vehicle.remove(veh_id)
 
+        # for veh_id in self.k.vehicle.get_ids():
```
Comment (kjang96) [inline flow/envs/multiagent/i210.py:363]: Remove

File: examples/exp_configs/non_rl/i210_subnetwork.py [hunk 11]
```diff
@@ -2,103 +2,184 @@
...
+            departSpeed=inflow_speed)
+
+    if PENETRATION_RATE > 0.0:
+        for lane in lane_list:
+            inflow.add(
+                veh_type="av",
+                edge=highway_start_edge,
+                vehs_per_hour=int(inflow_rate * PENETRATION_RATE),
+                departLane=lane,
+                departSpeed=inflow_speed)
```
[the remark this answers is not in the document]
Comment (eugenevinitsky) [reply examples/exp_configs/non_rl/i210_subnetwork.py:171]: Agreed but want this to land faster

File: examples/exp_configs/templates/sumo/i210_with_ghost_cell_with_downstream.xml [hunk 12]
```diff
@@ -3501,11 +3501,11 @@
...
-        <lane id="ghost0_0" index="0" speed="13.89" length="573.08" shape="-17.02,1355.30 478.76,1067.86"/>
-        <lane id="ghost0_1" index="1" speed="13.89" length="573.08" shape="-15.36,1358.16 480.42,1070.71"/>
-        <lane id="ghost0_2" index="2" speed="13.89" length="573.08" shape="-13.71,1361.01 482.07,1073.56"/>
-        <lane id="ghost0_3" index="3" speed="13.89" length="573.08" shape="-12.05,1363.86 483.73,1076.42"/>
-        <lane id="ghost0_4" index="4" speed="13.89" length="573.08" shape="-10.40,1366.72 485.38,1079.27"/>
+        <lane id="ghost0_0" index="0" speed="30" length="573.08" shape="-17.02,1355.30 478.76,1067.86"/>
+        <lane id="ghost0_1" index="1" speed="30" length="573.08" shape="-15.36,1358.16 480.42,1070.71"/>
+        <lane id="ghost0_2" index="2" speed="30" length="573.08" shape="-13.71,1361.01 482.07,1073.56"/>
+        <lane id="ghost0_3" index="3" speed="30" length="573.08" shape="-12.05,1363.86 483.73,1076.42"/>
+        <lane id="ghost0_4" index="4" speed="30" length="573.08" shape="-10.40,1366.72 485.38,1079.27"/>
```
[the remark this answers is not in the document]
Comment (eugenevinitsky) [reply examples/exp_configs/templates/sumo/i210_with_ghost_cell_with_downstream.xml:3508]: George said to.

[the remark this answers is not in the document]
Comment (eugenevinitsky) [reply examples/exp_configs/templates/sumo/i210_with_ghost_cell_with_downstream.xml:3508]: Haha

File: flow/visualize/visualizer_rllib.py (hunk 5 above)
Comment (eugenevinitsky) [reply flow/visualize/visualizer_rllib.py:173]: Nah, you want to turn it off at test time. 'll leave a comment

File: examples/train.py [hunk 13]
```diff
@@ -172,65 +174,128 @@ def setup_exps_rllib(flow_params,
...
-        speed = np.mean([speed for speed in env.k.vehicle.get_speed(env.k.vehicle.get_ids()) if speed >= 0])
+        if hasattr(env, 'invalid_control_edges'):
+            veh_ids = [veh_id for veh_id in env.k.vehicle.get_ids() if (env.k.vehicle.get_speed(veh_id) >= 0
+                                                                        and env.k.vehicle.get_edge(veh_id)
+                                                                        not in env.invalid_control_edges)]
+            rl_ids = [veh_id for veh_id in env.k.vehicle.get_rl_ids() if (env.k.vehicle.get_speed(veh_id) >= 0
+                                                                        and env.k.vehicle.get_edge(veh_id)
+                                                                        not in env.invalid_control_edges)]
+        else:
+            veh_ids = [veh_id for veh_id in env.k.vehicle.get_ids() if env.k.vehicle.get_speed(veh_id) >= 0]
```
[the remark this answers is not in the document]
Comment (eugenevinitsky) [reply examples/train.py:266]: Unfortunately doesn't quite work here

File: flow/envs/test.py (hunk 4 above)
Comment (eugenevinitsky) [reply flow/envs/test.py]: Deleted.

File: flow/controllers/base_controller.py [hunk 14]
```diff
@@ -114,6 +114,8 @@ def get_action(self, env):
         # time step
         if accel is None:
             return None
+        else:
+            accel = min(max(accel, -self.max_deaccel), self.max_accel)
```
Comment (liljonnystyle) [inline flow/controllers/base_controller.py:118]: this is already added into the Flow failsafes, so if you want it, use `fail_safe='feasible_accel'`

Comment (eugenevinitsky) [reply flow/controllers/base_controller.py:118]: thanks! removing

File: flow/envs/multiagent/__init__.py [hunk 15]
```diff
@@ -10,7 +10,8 @@
 from flow.envs.multiagent.traffic_light_grid import MultiTrafficLightGridPOEnv
 from flow.envs.multiagent.highway import MultiAgentHighwayPOEnv
 from flow.envs.multiagent.merge import MultiAgentMergePOEnv
-from flow.envs.multiagent.i210 import I210MultiEnv, MultiStraightRoad
+from flow.envs.multiagent.i210 import I210MultiEnv, MultiStraightRoad, I210MADDPGMultiEnv, MultiStraightRoadMADDPG
```
Comment (kjang96) [inline flow/envs/multiagent/__init__.py:13]: I210MADDPGMultiEnv import doesn't work. Try running `python train.py multiagent_i210 test` to verify

File: flow/algorithms/custom_ppo.py [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+
+import logging
+
+import numpy as np
+import ray
+from ray.rllib.agents.ppo.ppo import PPOTrainer
+from ray.rllib.evaluation.postprocessing import compute_advantages, \
+    Postprocessing
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.policy.tf_policy import LearningRateSchedule, \
```
Comment (kjang96) [inline flow/algorithms/custom_ppo.py:14]: the ACTION_LOGP import fails for me

File: examples/exp_configs/rl/multiagent/multiagent_i210.py [hunk 17]
```diff
@@ -49,84 +62,180 @@
...
+        inflow.add(
+            veh_type="human",
+            edge=highway_start_edge,
+            vehs_per_hour=int(inflow_rate * (1 - PENETRATION_RATE)),
+            departLane=lane,
+            departSpeed=inflow_speed)
+
+    inflow.add(
+        veh_type="human",
+        edge="27414345",
```
Comment (kjang96) [inline examples/exp_configs/rl/multiagent/multiagent_i210.py:146]: When I set `WANT_GHOST_CELL = True` and `ON_RAMP = True`, I get a KeyError. KeyError: '27414345'. I guess the edges are off? Script was `python train.py multiagent_i210 test`

Comment (kjang96) [reply examples/exp_configs/rl/multiagent/multiagent_i210.py:146]: Also happens for `WANT_GHOST_CELL = False` and `ON_RAMP = True`

Comment (kjang96) [reply examples/exp_configs/rl/multiagent/multiagent_i210.py:146]: False False is fine. Guess it's an issue with the edges in ON_RAMP

Comment (eugenevinitsky) [reply examples/exp_configs/rl/multiagent/multiagent_i210.py:146]: resolved

**Pull request merged**